### ISSUE TITLE
x/tools/go/analysis/passes/sigchanyzer: alteration of the AST

### ISSUE THREAD

**Issue by ldez** [golang/go#46129]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.16.4 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

yes, but the issue is not related to Go itself.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/ldez/.cache/go-build"
GOENV="/home/ldez/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/ldez/sources/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/ldez/sources/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/ldez/.gvm/gos/go1.16.4"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/ldez/.gvm/gos/go1.16.4/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16.4"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/ldez/sources/go/src/golang.org/x/tools/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build184925693=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?

I use go vet passes into golangci-lint.
`sigchanyzer` produces an unexpected behavior with the `buildssa.Analyzer`, I think it's because it alters the AST.

The problem comes from the following lines:
https://github.com/golang/tools/blob/be4aaae4cf865bd1e65ae3a22df28c1f47b6ccfd/go/analysis/passes/sigchanyzer/sigchanyzer.go#L62-L65

More details here: https://github.com/golangci/golangci-lint/issues/1973
Steps to reproduce: https://github.com/golangci/golangci-lint/issues/1973#issuecomment-839848505

### What did you expect to see?

No impact on the `buildssa.Analyzer`.

### What did you see instead?

Impact on the `buildssa.Analyzer`.

**Comment by bcmills** [comment]
`sigchanyzer` is new as of Go 1.17. Does this affect other analyses in `cmd/vet`?

(CC @cuonglm @odeke-em @stamblerre)

**Comment by ldez** [comment]
I created a potential fix: https://go-review.googlesource.com/c/tools/+/319211

**Comment by cuonglm** [comment]
@bcmills No, the other two passes that require `buildssa.Analyzer` are `nilness` and `unusedwrite`.

**Comment by cuonglm** [comment]
It's unclear to me though why modifying AST in one pass (sigchanyzer) will affect the other `buildssa`? Is there any documentation for this behavior?

**Comment by timothy-king** [comment]
> sigchanyzer is new as of Go 1.17. Does this affect other analyses in cmd/vet?
Yes.

tl;dr Don't modify the ast from an Analyzer.*

Modifying anything reachable from a Pass.Files during an Analyzer.Run(*Pass), let's call this A, races with every Analyzer B where A and B are not related by a Analyzer.Requires field. These are not working on unique copies of the AST for each pass but a shared one for each package.Package loaded.

Definition of action.pkg:
https://github.com/golang/tools/blob/v0.1.1/go/analysis/internal/checker/checker.go#L534
Example place analysis forwards action.pkg.Syntax: https://github.com/golang/tools/blob/v0.1.1/go/analysis/internal/checker/checker.go#L640
Where inspector gets the ast from pass.Files: https://github.com/golang/tools/blob/v0.1.1/go/analysis/passes/inspect/inspect.go#L48

There is nothing special about buildssa it just happens to be independently runnable from sigchanyzer.

* Unless you know all of the Analyzer being run and are in a Requires relationship to all of them.
